plugins/vpc-shared-eni/network/bridge_windows.go

```diff
@@ -273,29 +289,37 @@ func (nb *BridgeBuilder) FindOrCreateEndpoint(nw *Network, ep *Endpoint) error {
 
 // DeleteEndpoint deletes an existing HNS endpoint.
 func (nb *BridgeBuilder) DeleteEndpoint(nw *Network, ep *Endpoint) error {
-	// Query the infrastructure container ID.
-	isInfraContainer, infraContainerID, err := nb.getInfraContainerID(ep)
-	if err != nil {
-		return err
-	}
+	// Query the namespace identifier.
+	nsType, namespaceIdentifier := nb.getNamespaceIdentifier(ep)
 
 	// Find the HNS endpoint ID.
-	endpointName := nb.generateHNSEndpointName(ep, infraContainerID)
+	endpointName := nb.generateHNSEndpointName(ep, namespaceIdentifier)
 	hnsEndpoint, err := hcsshim.GetHNSEndpointByName(endpointName)
 	if err != nil {
 		return err
 	}
 
 	// Detach the HNS endpoint from the container's network namespace.
 	log.Infof("Detaching HNS endpoint %s from container %s netns.", hnsEndpoint.Id, ep.ContainerID)
-	err = hcsshim.HotDetachEndpoint(ep.ContainerID, hnsEndpoint.Id)
-	if err != nil && err != hcsshim.ErrComputeSystemDoesNotExist {
-		return err
-	}
+	if nsType == hcsNamespace {
+		// Detach the HNS endpoint from the namespace, if we can.
+		// HCN Namespace and HNS Endpoint have a 1-1 relationship, therefore,
+		// even if detachment of endpoint from namespace fails, we can still proceed to delete it.
+		err = hcn.RemoveNamespaceEndpoint(namespaceIdentifier, hnsEndpoint.Id)
+		if err != nil {
+			log.Errorf("Failed to detach endpoint, ignoring: %v", err)
+		}
+	} else {
+		err = hcsshim.HotDetachEndpoint(ep.ContainerID, hnsEndpoint.Id)
+		if err != nil && err != hcsshim.ErrComputeSystemDoesNotExist {
+			return err
+		}
 
-	// The rest of the delete logic applies to infrastructure container only.
-	if !isInfraContainer {
-		return nil
+		// The rest of the delete logic applies to infrastructure container only.
+		if nsType == nonInfraContainerNS {
+			// For non-infra containers, the network must not be deleted.
+			return nil
+		}
 	}
 
 	// Delete the HNS endpoint.
```

Comment:
Wont this fail if the `hcn.RemoveNamespaceEndpoint` call succeeds though. Should we handle the `notfound` error?

Reply:
`hcn.RemoveNamespaceEndpoint` only removes the endpoint ID from the HCN Namespace, i.e. detaches the endpoint from the HCN namespace. The endpoint would still exist. The last call is to delete the endpoint itself.

Reply:
Oh I see. Ok that makes sense then. Once we move fully to hcn this will get changed as well but that is for a future date. The unified path here works